pkg/worker/builder.go

```diff
@@ -59,42 +59,35 @@ func (b builder) Build(ctx context.Context) (Result, error) {
 		result.Status = workermodel.StatusNone
 		return result, nil
 	}
+	anyPreviousStageHasFailed := false
 	for _, stageRunner := range b.stageRunners {
+		stagesDone++
+		stage := stageRunner.Stage()
+		if stage.ShouldSkip(anyPreviousStageHasFailed) {
+			logSkippedStage(stage, stagesDone, stagesCount)
+			continue
+		}
 		log.Info().
 			WithStringf("stages", "%d/%d", stagesDone, stagesCount).
-			WithString("stage", stageRunner.Stage().Name).
+			WithString("stage", stage.Name).
 			Message("Starting stage.")
 		res := stageRunner.RunStage(ctx)
 		result.Stages = append(result.Stages, res)
-		stagesDone++
 		if res.Status != workermodel.StatusSuccess {
-			var failed []string
-			var cancelled []string
-			for _, stepRes := range res.Steps {
-				if stepRes.Status == workermodel.StatusFailed {
-					failed = append(failed, stepRes.Name)
-				} else if stepRes.Status == workermodel.StatusCancelled {
-					cancelled = append(cancelled, stepRes.Name)
-				}
+			logFailedStage(res, stagesDone, stagesCount)
+			if !anyPreviousStageHasFailed {
+				log.Debug().Message("Skipping `run-if: success` stages from now on.")
+				anyPreviousStageHasFailed = true
 			}
-			log.Warn().
-				WithStringf("stages", "%d/%d", stagesDone, stagesCount).
-				WithString("stage", res.Name).
-				WithDuration("dur", res.Duration.Truncate(time.Second)).
-				WithStringer("status", res.Status).
-				WithString("failed", strings.Join(failed, ",")).
-				WithString("cancelled", strings.Join(cancelled, ",")).
-				Message("Failed stage. Skipping any further stages.")
 			result.Status = res.Status
-			break
+			continue
 		}
-		log.Info().
-			WithStringf("stages", "%d/%d", stagesDone, stagesCount).
-			WithString("stage", res.Name).
-			WithDuration("dur", res.Duration.Truncate(time.Second)).
-			Message("Done with stage.")
+		logSuccessfulStage(res, stagesDone, stagesCount)
 		result.Status = workermodel.StatusSuccess
 	}
+	if anyPreviousStageHasFailed {
+		result.Status = workermodel.StatusFailed
+	}
```

Comment:
Without this change, in the case where we have a wharf-ci file like:

S
F

Where `S` is `run-if: success` and `F` is `run-if: fail`, the build could be reported as successful even if `S` failed.

I think there needs to be an option for exempting a stage from the final build result status report, as some conditional stages might be used for logging or clean up of some sort.

How often would such a feature be necessary though? Is it something that is common?

Reply:
In Azure Pipelines if any stage failed but it's "caught", the status will be orange/PartiallySucceeded. So we could consider adding a new status like that.

But I think for now, it's okay with this change so that if any stage has failed, the result is failed.

Reply:
That sounds good too.
I created an issue on it here https://github.com/iver-wharf/wharf-cmd/issues/196.